tests/core/test_dataset.py

```diff
@@ -0,0 +1,180 @@
+    assert len(batch_request.requests) == 2
+    assert batch_request.requests[0].logical_series_rid == "ch-1"
+    assert batch_request.requests[1].logical_series_rid == "ch-2"
+    mock_available_units.assert_called_once()
+
+
+@patch("nominal.core.dataset._available_units", return_value=UNITS)
+@patch.object(Dataset, "get_channels")
+def test_set_channel_units_invalid_unit(
+    mock_get_channels: MagicMock, mock_available_units: MagicMock, mock_dataset: Dataset
+):
+    mock_get_channels.return_value = [
+        Channel(
+            rid="ch-1",
+            name="channel1",
+            data_source="ds-1",
+            data_type="float",
+            unit=None,
+            description="Test Channel 1",
+            _clients=mock_dataset._clients,
+        ),
+    ]
+
+    invalid_channels_to_units = {"channel1": "invalid_unit"}
+
+    with pytest.raises(ValueError, match="Provided unit 'invalid_unit' for channel 'channel1'"):
+        mock_dataset.set_channel_units(invalid_channels_to_units)
+    mock_available_units.assert_called_once()
+
+
+@pytest.mark.parametrize("validate", [True, False])
+@patch("nominal.core.dataset._available_units", return_value=UNITS)
+@patch.object(Dataset, "get_channels")
+def test_set_channel_units_no_channel_data(
+    mock_get_channels: MagicMock, mock_available_units: MagicMock, mock_dataset: Dataset, validate: bool
+):
+    mock_get_channels.return_value = []
+    channels_to_units = {"channel1": "kg"}
+    call = functools.partial(mock_dataset.set_channel_units, channels_to_units, validate_schema=validate)
+    if validate:
+        with pytest.raises(ValueError, match="Unable to set unit for channel1 to kg: no data uploaded for channel"):
+            call()
+    else:
+        call()
+        mock_dataset._clients.logical_series.batch_update_logical_series.assert_not_called()
+    mock_available_units.assert_called_once()
+
+
+@patch("time.sleep", return_value=None)
+def test_poll_until_ingestion_completed_success(mock_sleep: MagicMock, mock_dataset: Dataset):
+    mock_dataset._clients.catalog.get_ingest_progress_v2.return_value = MagicMock(
+        ingest_status=MagicMock(type="success")
+    )
+
+    mock_dataset.poll_until_ingestion_completed(interval=timedelta(seconds=1))
+
+    mock_dataset._clients.catalog.get_ingest_progress_v2.assert_called()
+
+
+@patch("time.sleep", return_value=None)
+def test_poll_until_ingestion_completed_in_progress(mock_sleep: MagicMock, mock_dataset: Dataset):
+    mock_dataset._clients.catalog.get_ingest_progress_v2.side_effect = [
+        MagicMock(ingest_status=MagicMock(type="inProgress")),
+        MagicMock(ingest_status=MagicMock(type="inProgress")),
+        MagicMock(ingest_status=MagicMock(type="success")),
+    ]
+
+    mock_dataset.poll_until_ingestion_completed(interval=timedelta(seconds=1))
+
+    assert mock_dataset._clients.catalog.get_ingest_progress_v2.call_count == 3
+    assert mock_sleep.call_count == 2
```

Comment:
ah I see. I dunno if I'd test if sleep is called in particular since it's kinda an impl detail (for example could use a condition variable or event instead)

Reply:
eh but there's no other way to test the polling frequency and it's trivial to change the test if it becomes an event or something instead